### PR TITLE
Add stable diffusion tflite inference pipeline

### PR DESCRIPTION
**Pull request by yichunk**
Add stable diffusion tflite inference pipeline and README for running the pipeline.

BUG=b/311216181

### CODE REVIEW

File: ai_edge_torch/generative/examples/stable_diffusion/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Stable Diffusion through TFLite
+This example show how to use Generative API to convert a Pytorch Stable Diffusion v1.5 model to TFLite model, and run the image generation inference.
```
Comment (haozha111) [inline ai_edge_torch/generative/examples/stable_diffusion/README.md:2]: the Edge Generative API?

File: ai_edge_torch/generative/examples/stable_diffusion/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+from ai_edge_torch.generative.examples.stable_diffusion.tokenizer import Tokenizer  # NOQA
+import ai_edge_torch.generative.examples.stable_diffusion.util as util
+from ai_edge_torch.model import TfLiteModel
+
+
+class StableDiffusion:
+
+  def __init__(
+      self,
+      *,
```
Comment (haozha111) [inline ai_edge_torch/generative/examples/stable_diffusion/pipeline.py:34]: what's this for?

File: ai_edge_torch/generative/examples/stable_diffusion/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  ):
+    self.tokenizer = Tokenizer(tokenizer_vocab_dir)
+    self.clip = TfLiteModel.load(clip_ckpt)
+    self.decoder = TfLiteModel.load(decoder_ckpt)
+    self.diffusion = TfLiteModel.load(diffusion_ckpt)
+    if encoder_ckpt is not None:
+      self.encoder = TfLiteModel.load(encoder_ckpt)
+
+
+def run_tflite_pipeline(
```
Comment (haozha111) [inline ai_edge_torch/generative/examples/stable_diffusion/pipeline.py:49]: can this python pipeline easily map to MediaPipe in the future?

File: ai_edge_torch/generative/examples/stable_diffusion/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  if not os.path.exists(output_path):
+    Path(output_path).parent.mkdir(parents=True, exist_ok=True)
+  Image.fromarray(images[0].astype(np.uint8)).save(output_path)
+
+
+if __name__ == '__main__':
+  run_tflite_pipeline(
+      StableDiffusion(
+          tokenizer_vocab_dir=os.path.join(
+              Path.home(), 'Downloads/stable_diffusion_data'
```
Comment (haozha111) [inline ai_edge_torch/generative/examples/stable_diffusion/pipeline.py:173]: can you allow user to provide a custom path ?

File: ai_edge_torch/generative/examples/stable_diffusion/tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def pairwise(seq):
+  a = iter(seq)
+  b = iter(seq)
+  next(b)
+  return zip(a, b)
+
+
+class Tokenizer:
```
Comment (haozha111) [inline ai_edge_torch/generative/examples/stable_diffusion/tokenizer.py:45]: is this tokenizer also compatible with MP/BYOM's tokenizer?

File: ai_edge_torch/generative/examples/stable_diffusion/util.py [hunk 6]
```diff
@@ -44,9 +44,15 @@ def get_file_path(filename, url=None):
 
 def move_channel(image, to):
   if to == "first":
-    return image.permute(0, 3, 1, 2)  # (N, H, W, C) -> (N, C, H, W)
+    if isinstance(image, torch.Tensor):
```
Comment (haozha111) [inline ai_edge_torch/generative/examples/stable_diffusion/util.py:47]: can you tell me why we are doing this?

File: ai_edge_torch/generative/examples/stable_diffusion/pipeline.py (hunk 3 above)
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/pipeline.py:49]: Should not be a heavy work, given the stable diffusion is implemented with mediapipe before. We need to map the pipeline to the existing mediapipe graph.

File: ai_edge_torch/generative/examples/stable_diffusion/pipeline.py (hunk 2 above)
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/pipeline.py:34]: This requires the following parameters to be specified by key words.

File: ai_edge_torch/generative/examples/stable_diffusion/tokenizer.py (hunk 5 above)
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/tokenizer.py:45]: There is a counterpart CLIP tokenizer in mediapipe.

File: ai_edge_torch/generative/examples/stable_diffusion/util.py (hunk 6 above)
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/util.py:47]: The util was created for torch.Tensor.  The update makes it compatible with both torch.Tensor and numpy.ndarray

File: ai_edge_torch/generative/examples/stable_diffusion/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/README.md]: Done

[the remark this answers is not in the document]
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/README.md]: Done

Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/README.md]: Done

File: ai_edge_torch/generative/examples/stable_diffusion/README.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+## Convert Pytorch to TFLite model
+1. Download pytorch stable diffusion model [stable-diffusion-pytorch](https://github.com/kjsman/stable-diffusion-pytorch)
```
[the remark this answers is not in the document]
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/README.md]: Done

File: ai_edge_torch/generative/examples/stable_diffusion/README.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+1. Unzip the downloaded model weight into `$HOME/Downloads/stable_diffusion_data`
+1. Run `convert_to_tflite.py`. This will convert the pytorch models into tflite models. The stable diffusion model has four components: CLIP (text embedding), encoder, diffusion and decoder models. Each component is converted to a single tflite model file.
```
[the remark this answers is not in the document]
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/README.md]: Done

File: ai_edge_torch/generative/examples/stable_diffusion/pipeline.py (hunk 4 above)
Comment (yichunk) [reply ai_edge_torch/generative/examples/stable_diffusion/pipeline.py]: Done. Update the script arguments.

**Review by haozha111 — APPROVED** [review]

**Pull request merged**